hs/src/LedgerState.hs

```diff
@@ -265,52 +274,56 @@ applyTxBody :: LedgerState -> Tx -> LedgerState
 applyTxBody ls tx = ls { getUtxo = newUTxOs }
   where newUTxOs = txins tx </| getUtxo ls `union` txouts tx
 
--- |Apply a certificate as a state transition function on the ledger state.
-applyCert :: Cert -> LedgerState -> LedgerState
-applyCert (RegKey key) ls@(LedgerState _ ds _) = ls
+-- |Apply a delegation certificate as a state transition function on the ledger state.
+applyDCert :: Slot -> DCert -> LedgerState -> LedgerState
+applyDCert slot (RegKey key) ls@(LedgerState _ ds _) = ls
   { getDelegationState = ds
-    { getStKeys = Set.insert hk_sk (getStKeys ds)
-    , getAccounts = Map.insert hk_sk (Coin 0) (getAccounts ds)}
+    { getStKeys = Map.insert hksk slot (getStKeys ds)
+    , getAccounts = Map.insert (RewardAcnt hksk) (Coin 0) (getAccounts ds)}
   }
-  where hk_sk = hashKey key
+  where hksk = hashKey key
 
-applyCert (DeRegKey key) ls@(LedgerState _ ds _) = ls
+applyDCert _ (DeRegKey key) ls@(LedgerState _ ds _) = ls
   { getDelegationState = ds
-    { getStKeys = Set.delete hk_sk (getStKeys ds)
-    , getAccounts = Map.delete hk_sk (getAccounts ds)
-    , getDelegations = Map.delete hk_sk (getDelegations ds) }
+    { getStKeys = Map.delete hksk (getStKeys ds)
+    , getAccounts = Map.delete (RewardAcnt hksk) (getAccounts ds)
+    , getDelegations = Map.delete hksk (getDelegations ds) }
   }
-  where hk_sk = hashKey key
+  where hksk = hashKey key
 
 -- TODO do we also have to check hashKey target?
-applyCert (Delegate (Delegation source target)) ls@(LedgerState _ ds _) = ls
+applyDCert _ (Delegate (Delegation source target)) ls@(LedgerState _ ds _) = ls
   {getDelegationState = ds
     { getDelegations =
         Map.insert (hashKey source) (hashKey target) (getDelegations ds)}
   }
 
 -- TODO what happens if there's already a pool registered with that key?
-applyCert (RegPool sp) ls@(LedgerState _ ds _) = ls
+applyDCert slot (RegPool sp) ls@(LedgerState _ ds _) = ls
   { getDelegationState = ds
-    { getStPools = Set.insert sp (getStPools ds)
+    { getStPools = Map.insert hsk (sp, slot') pools
     , getRetiring = Map.delete hsk (getRetiring ds)}}
   where hsk = hashKey $ poolPubKey sp
+        pools = getStPools ds
+        slot' = case Map.lookup hsk pools of
+                  Just (_, s) -> s
+                  Nothing -> slot
```

Comment:
I see it's done in the same sense in the rule. But aren't we effectively using the "update the existing map entry" backwards here ? How about using a `\unionDoNotOverride` or similar instead? (I know it's not in the scope of this PR, but I'd say the rule would be less complex then.

Reply:
The trouble is that the value is a pair, one part of which we override and the other which we don't. maybe we should just have two maps, that would clean up the notation elsewhere as well.